Content.IntegrationTests/Tests/Lathe/LatheTest.cs

```diff
@@ -0,0 +1,98 @@
+        var server = pair.Server;
+
+        var mapData = await pair.CreateTestMap();
+
+        var entMan = server.EntMan;
+        var protoMan = server.ProtoMan;
+        var compFactory = server.ResolveDependency<IComponentFactory>();
+        var materialStorageSystem = server.System<SharedMaterialStorageSystem>();
+        var whitelistSystem = server.System<EntityWhitelistSystem>();
+
+        await server.WaitAssertion(() =>
+        {
+            // Find all the lathes
+            var latheProtos = protoMan.EnumeratePrototypes<EntityPrototype>()
+                .Where(p => !p.Abstract)
+                .Where(p => !pair.IsTestPrototype(p))
+                .Where(p => p.HasComponent<LatheComponent>());
+
+            // Find every EntityPrototype that can be inserted into a MaterialStorage
+            var materialEntityProtos = protoMan.EnumeratePrototypes<EntityPrototype>()
+                .Where(p => !p.Abstract)
+                .Where(p => !pair.IsTestPrototype(p))
+                .Where(p => p.HasComponent<PhysicalCompositionComponent>());
+
+            // Spawn all of the above material EntityPrototypes - we need actual entities to do whitelist checks
+            var materialEntities = new List<EntityUid>(materialEntityProtos.Count());
+            foreach (var materialEntityProto in materialEntityProtos)
+            {
+                materialEntities.Add(entMan.SpawnEntity(materialEntityProto.ID, mapData.GridCoords));
+            }
+
+            Assert.Multiple(() =>
+            {
+                // Check each lathe individually
+                foreach (var latheProto in latheProtos)
+                {
+                    if (!latheProto.TryGetComponent<LatheComponent>(out var latheComp, compFactory))
+                        continue;
+
+                    if (!latheProto.TryGetComponent<MaterialStorageComponent>(out var storageComp, compFactory))
+                        continue;
+
+                    // Test which material-containing entities are accepted by this lathe
+                    var acceptedMaterials = new HashSet<ProtoId<MaterialPrototype>>();
+                    foreach (var materialEntity in materialEntities)
+                    {
+                        Assert.That(entMan.TryGetComponent<PhysicalCompositionComponent>(materialEntity, out var compositionComponent));
+                        if (whitelistSystem.IsWhitelistFail(storageComp.Whitelist, materialEntity))
+                            continue;
+
+                        // Mark the lathe as accepting each material in the entity
+                        foreach (var (material, _) in compositionComponent.MaterialComposition)
+                        {
+                            acceptedMaterials.Add(material);
+                        }
+                    }
+
+                    // Check each recipe assigned to this lathe
+                    foreach (var recipeId in latheComp.StaticRecipes)
```

Comment:
The test should also cover the `DynamicRecipes` which are unlocked with research.

Reply:
Oh, and the additional recipes in the `EmagLatheRecipesComponent` as well.